Skipping hidden files compilation for script service

fforbeck opened the pull request:
If hidden files are detected during the ScriptService initialization, they will be ignored and a warning message will be logged.

Closes #15269

clintongormley commented:
@rjernst could you review please?

jasontedor commented on core/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -527,6 +528,10 @@ public void onFileInit(Path file) {
             if (logger.isTraceEnabled()) {
                 logger.trace("Loading script file : [{}]", file);
             }
+            if (FileSystemUtils.isHidden(file)) {
```
I think this should go before the previous check because otherwise a trace log will read that the script file is being loaded, and then that it is being skipped.

jasontedor commented on core/src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -527,6 +528,10 @@ public void onFileInit(Path file) {
                 logger.trace("Loading script file : [{}]", file);
             }
+            if (FileSystemUtils.isHidden(file)) {
+                logger.warn("--- Hidden file skipped : [{}]", file);
```
I'm not sure if warn is the right level here, maybe debug. Can you remove the leading "--- " and the word "script" to the log message somewhere?

nik9000 replied:
I agree with @jasontedor .  I think `debug` is fine here. If I had to guess this is a leftover from debugging the implementation. I know I tend to use stuff like `---` or `asdfewfca` to mark any temporary logs I add.

jasontedor commented on core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java:
```diff
@@ -148,6 +148,39 @@ public void testScriptsWithoutExtensions() throws IOException {
     }
 
+    public void testHiddenFileSkipped() throws IOException {
+        ContextAndHeaderHolder contextAndHeaders = new ContextAndHeaderHolder();
```
This is gone from master now, you'll have to integrate master and rework.

jasontedor commented on core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java:
```diff
@@ -148,6 +148,39 @@ public void testScriptsWithoutExtensions() throws IOException {
+        logger.info("--> delete hidden file");
+        Files.delete(testFileHidden);
+
+        logger.info("--> delete test file");
```
There are several logging statements like this one and I'm not sure what they are adding to the test? If the test fails, a stack trace will give a pretty clear indication how far the test had progressed.

jasontedor commented on core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java:
```diff
@@ -148,6 +148,39 @@ public void testScriptsWithoutExtensions() throws IOException {
+
+        try {
+            logger.info("--> verify if hidden_file was skipped");
+            scriptService.compile(new Script("hidden_file", ScriptType.FILE, "test", null),
```
This test is not really testing what we want to be testing here. The reason that it's not is because the cache key for a file named ".hidden_file" is not "hidden_file", but rather it is "". A file named ".hidden_file" never would have been processed by the compilation engine because it doesn't have an extension. So this will ultimately throw, but not for the right reason.

fforbeck replied:
Hi @jasontedor, thanks for reviewing. 
I have fixed the other points, but only this one I am not sure what to do.
In this case then, would be better to use the key as `""`?

jasontedor replied:
I don't think so, that's relying too much on an internal implementation detail (that there is a cache, that file scripts are compiled and put into the static cache, that the key is the prefix of the filename, etc.). The purpose behind the PR is to get the script service to ignore hidden files, and that is what needs to be tested. I haven't looked too closely, but I suspect that you're going to have to hook into the script service or maybe the resource watcher and possibly refactor a little bit to expose the pieces needed to in fact make this assertion. Let me know if that's enough to get you started, I'm happy to take a closer look if needed. :)

fforbeck replied:
Alright! Yep, I think I can find a way out from this point.
I let you know. Thanks a lot!

fforbeck replied:
Hi @jasontedor, 
I was wondering if we could just change the method:
`private Tuple<String, String> scriptNameExt(Path file)` from `ScriptChangesListener` to ignore any kind of file that does not have a name. I mean, hidden files. 

It starts with `.` but the method allows you to retrieve an empty script name for this case. If we just use the `if (extIndex > 0)` we would ignore this kind of file. In addition I would move the method to the parent class and drop the access modifier to create the unit test. We would not need to check if the file is a hidden file or not.
Please, let me know what do you think. 
Thanks

rjernst commented:
@fforbeck Thanks for the update. In the future, you should post a message after you push new commits, otherwise we don't know they exist without checking in on the PR once in a while. :)

Changing the return value of `getScriptNameExt` seems ok. However, it looks like you moved it out of the ScriptChangesListener just so it can be tested. I think it would be better to keep it where it was, and write the test like other existing tests of file based scripts. See for example `ScriptServiceTests.testScriptsWithoutExtensions()`. While the test you added checks `getScriptNameExt` works correctly, it does not check the scripts are actually skipped correctly when found on disk. So I think a test actually using the `resourceWatcherService` in `ScriptServiceTests` would better ensure we don't accidentally break skipping over hidden files in future refactorings.

fforbeck commented:
Hi @rjernst thanks for your reply. Okay, next time I will make sure to post a message.

Regarding the test, in my first PR I had added the test to check if the file was properly skipped using the `resourceWatcherService`. 

@jasontedor pointed out that it wasn't actually testing the change because we will never have a cached key named "hidden_file" if the script file is called `.hidden_file`. The cached key would be an empty string, because the file has no name and the test case was trying to retrieve the file by `hidden_file` cached key.

Then I decided to move the `getScriptNameExt` the method out of the listener class and drop the modifier to write the test. In addition to hidden files, I got another case which we have an issue -
if the file does not have any extension.

So, I have updated the code to test the `getScriptNameExt` independently with valid and invalid file names, and added back the test for the `ScriptChangesListener` to check if the hidden script file is skipped.

Please let me know what you think.

Thank you!

rjernst commented:
@fforbeck I wasn't suggesting a test like you had before. I agree with @jasontedor that checking for non-existence of a script with an empty name isn't really testing anything. However, I think there a test which ensures an error does not occur when an arbitrary dot file exists, and that other scripts are still loaded, would be good. As I mentioned in my previous comment, your tests right now are not actually testing that the ScriptService works, only that this method (which really should stay internal to the file change listener) works by itself.

fforbeck commented:
@rjernst 
Alright! Got it! I have added the test to load scripts even if dot files are found and I moved that method to the inner class as you recommended. Could you review it please?

Thanks a lot!

rjernst commented:
Looks good, I merged, thanks @fforbeck!
